code/blake2/Hacl.Impl.Blake2.Generic.fst

```diff
@@ -549,7 +549,221 @@ let blake2_init_st  (al:Spec.alg) (ms:m_spec) =
+      as_seq h b == Seq.create 8 (u64 0)
+    )
+    (ensures fun h0 _ h1 ->
+      modifies (loc b) h0 h1 /\
+      as_seq h1 b == Spec.serialize_blake2_params
+        (Spec.set_key_length (Spec.set_digest_length (blake2_params_v h0 p) (v nn)) (v kk)))
+  = let h0 = ST.get () in
+    [@inline_let]
+    let kk_shift_8 = shift_left (to_u64 kk) (size 8) in
+    [@inline_let]
+    let fanout_shift_16 = shift_left (to_u64 p.fanout) (size 16) in
+    [@inline_let]
+    let depth_shift_24 = shift_left (to_u64 p.depth) (size 24) in
+    [@inline_let]
+    let leaf_length_shift_32 = shift_left (to_u64 p.leaf_length) (size 32) in
+    [@inline_let]
+    let v0 = (to_u64 nn) ^. kk_shift_8 ^. fanout_shift_16 ^. depth_shift_24 ^. leaf_length_shift_32 in
+    [@inline_let]
+    let xof_length_shift_32 = shift_left (to_u64 p.xof_length) (size 32) in
+    [@inline_let]
+    let v1 = (to_u64 p.node_offset) ^. xof_length_shift_32 in
+    [@inline_let]
+    let inner_length_shift_8 = shift_left (to_u64 p.inner_length) (size 8) in
+    [@inline_let]
+    let v2 = (to_u64 p.node_depth) ^. inner_length_shift_8 in
+
+    uints_from_bytes_le (sub b 4ul 2ul) p.salt;
+    uints_from_bytes_le (sub b 6ul 2ul) p.personal;
+
+    b.(0ul) <- v0;
+    b.(1ul) <- v1;
+    b.(2ul) <- v2;
+    b.(3ul) <- (u64 0);
+
+    let h1 = ST.get () in
+    let aux () : Lemma (as_seq h1 b `Seq.equal` Spec.serialize_blake2b_params
+        (Spec.set_key_length (Spec.set_digest_length (blake2_params_v h0 p) (v nn)) (v kk))) =
+      let open Lib.Sequence in
+      let open Lib.ByteSequence in
+      let s0 = (u64 (v nn)) ^.
+               (u64 (v kk) <<. (size 8)) ^.
+               (u64 (v p.fanout) <<. (size 16)) ^.
+               (u64 (v p.depth) <<. (size 24)) ^.
+               (u64 (v p.leaf_length) <<. (size 32)) in
+      let s1 = (u64 (v p.node_offset)) ^.
+               (u64 (v p.xof_length) <<. (size 32)) in
+      // The serialization corresponding to s2 contains node_depth and inner_length,
+      // followed by the 14 reserved bytes which always seem to be zeros, and can hence
+      // be ignored when building the corresponding uint64 using xor's
+      let s2 = (u64 (v p.node_depth)) ^.
+               (u64 (v p.inner_length) <<. (size 8)) in
+      // s3 corresponds to the remaining of the reserved bytes
+      let s3 = u64 0 in
+      let salt_u64: lseq uint64 2 = uints_from_bytes_le (as_seq h0 (get_salt p)) in
+      let s4 = salt_u64.[0] in
+      let s5 = salt_u64.[1] in
+      let personal_u64: lseq uint64 2 = uints_from_bytes_le (as_seq h0 (get_personal p)) in
+      let s6 = personal_u64.[0] in
+      let s7 = personal_u64.[1] in
+      [@inline_let]
+      let l = [s0; s1; s2; s3; s4; s5; s6; s7] in
+      assert_norm (List.Tot.length l == 8);
+
+      // There seems to be something not triggering with createL, requiring the
+      // following lemma calls, and assert_norms to relate List.index to the
```

Comment:
there's supposed to be helpers like seq_of_list that do this automatically

Reply:
I looked into this, but, as far as I understand (I might have missed a different lemma), the lemmas should be triggered automatically when using createL. See the comment below, I'm not sure why this is not the case here.